src/xenia/kernel/friend_presence_object_json.cc

```diff
@@ -0,0 +1,145 @@
+ * Copyright 2024 Xenia Emulator. All rights reserved.                        *
+ * Released under the BSD license - see LICENSE in the root for more details. *
+ ******************************************************************************
+ */
+
+#include <string>
+
+#include "xenia/base/string_util.h"
+#include "xenia/kernel/friend_presence_object_json.h"
+#include "xenia/kernel/util/net_utils.h"
+
+namespace xe {
+namespace kernel {
+FriendPresenceObjectJSON::FriendPresenceObjectJSON()
+    : xuid_(0),
+      gamertag_(""),
+      state_(0),
+      sessionId_(0),
+      title_id_(""),
+      state_change_time_(0),
+      rich_state_presence_size_(0),
+      rich_presence_(0) {}
+
+FriendPresenceObjectJSON::~FriendPresenceObjectJSON() {}
+
+bool FriendPresenceObjectJSON::Deserialize(const rapidjson::Value& obj) {
+  if (obj.HasMember("xuid")) {
+    const auto xuid_str = std::string(obj["xuid"].GetString());
+
+    if (!xuid_str.empty()) {
+      XUID(string_util::from_string<uint64_t>(xuid_str, true));
+    }
+  }
+
+  if (obj.HasMember("gamertag")) {
+    Gamertag(obj["gamertag"].GetString());
+  }
+
+  if (obj.HasMember("state")) {
+    State(obj["state"].GetUint());
+  }
+
+  if (obj.HasMember("sessionId")) {
+    const auto sessionId_str = std::string(obj["sessionId"].GetString());
+
+    if (!sessionId_str.empty()) {
+      SessionID(string_util::from_string<uint64_t>(sessionId_str, true));
+    }
+  }
+
+  if (obj.HasMember("titleId")) {
+    TitleID(obj["titleId"].GetString());
+  }
+
+  if (obj.HasMember("stateChangeTime")) {
+    StateChangeTime(obj["stateChangeTime"].GetUint64());
+  }
+
+  if (obj.HasMember("richPresenceStateSize")) {
+    RichStatePresenceSize(obj["richPresenceStateSize"].GetUint());
+  }
+
+  // char_16?
+  if (obj.HasMember("richPresence")) {
+    // RichPresence(obj["richPresence"].GetString());
+  }
+
+  return true;
+}
+
+bool FriendPresenceObjectJSON::Serialize(
+    rapidjson::PrettyWriter<rapidjson::StringBuffer>* writer) const {
+  // writer->StartObject();
+
+  // writer->String("xuid");
+  // writer->String(fmt::format("{:016X}", xuid_));
+
+  // writer->String("gamertag");
+  // writer->Uint(gamertag_);
+
+  // writer->String("state");
+  // writer->Uint(state_);
+
+  // writer->String("sessionId");
+  // writer->String(fmt::format("{:016X}", sessionId_));
+
+  // writer->String("titleId");
+  // writer->String(title_id_);
+
+  // writer->String("stateChangeTime");
+  // writer->Uint64(state_change_time_);
+
+  // writer->String("richPresenceStateSize");
+  // writer->Uint(rich_state_presence_size_);
+
+  //// Problem :/
+  // writer->String("richPresence");
+  //// writer->String(rich_presence_);
+
+  // writer->EndObject();
+
+  return false;
+}
+
+FriendsPresenceObjectJSON::FriendsPresenceObjectJSON() : xuids_(0) {}
+
+FriendsPresenceObjectJSON::~FriendsPresenceObjectJSON() {}
+
+bool FriendsPresenceObjectJSON::Deserialize(const rapidjson::Value& obj) {
+  if (obj.IsArray()) {
```

Comment:
Less nesting with negating this check, unless you want to add more types to check